### PR TITLE
[CLOUDTRUST-1157] Add user_id in events + allow generic HTTP reponse

### PR DESCRIPTION
**Pull request by fperot74**
Generic response let REST methods be able to choose a status code, set headers, return mime content, ...

### CODE REVIEW

File: database/events.go [hunk 1]
```diff
@@ -48,41 +96,54 @@ func NewEventsDBModule(db CloudtrustDB) EventsDBModule {
...
+			}
+		}
+		if additionalInfoBytes, err := json.Marshal(addNfo); err == nil && len(addNfo) > 0 {
+			additionalInfo = string(additionalInfoBytes)
+		}
+	}
 
 	//store the event in the DB
-	_, err := cm.db.Exec(insertEvent, auditTime, origin, realmName, agentUserID, agentUsername, agentRealmName, userID, username, ctEventType, kcEventType, kcOperationType, clientID, additionalInfo)
+	_, err := cm.db.Exec(insertEvent, auditTime, origin, checkNull(realmName), checkNull(agentUserID), checkNull(agentUsername),
```
Comment (bsoniam) [inline database/events.go]: Why do we need to check for empty string ?

Comment (fperot74) [reply database/events.go]: It would be better to have NULL in database rather than empty string.

File: http/handlers.go [hunk 2]
```diff
@@ -78,13 +131,12 @@ func EncodeReply(_ context.Context, w http.ResponseWriter, rep interface{}) erro
...
 
-	w.Header().Set("Content-Type", "application/json; charset=utf-8")
-	w.WriteHeader(http.StatusOK)
-
-	var json, err = json.MarshalIndent(rep, "", " ")
-
-	if err == nil {
-		w.Write(json)
+	switch e := rep.(type) {
+	case GenericResponse:
```
Comment (harture) [inline http/handlers.go]: GenericResponse is HTTP oriented. 
The intial purpose was to split it between an answer that can be translated into a specific kind of response, HTTP or gRPC, ... I know that we don't have it anymore but I think it was a good split of concerns.

File: http/handlers.go [hunk 3]
```diff
@@ -15,6 +15,59 @@ import (
...
+	MimeType string
+	Content  []byte
+}
+
+// GenericResponse for HTTP requests
+type GenericResponse struct {
+	StatusCode   int
+	Headers      map[string]string
+	MimeContent  *MimeContent
+	ExportToJSON interface{}
```
[the remark this answers is not in the document]
Comment (fperot74) [reply http/handlers.go]: As discussed together, renamed to JSONableResponse

File: database/events.go [hunk 4]
```diff
@@ -48,41 +96,54 @@ func NewEventsDBModule(db CloudtrustDB) EventsDBModule {
...
 	// kcOperationType - operation type of the event that comes from Keycloak
-	kcOperationType := m["kc_operation_type"]
+	kcOperationType := m[CtEventKcOperationType]
 	// Id of the client
-	clientID := m["client_id"]
+	clientID := m[CtEventClientID]
 	//additional_info - all the rest of the information from the event
-	additionalInfo := m["additional_info"]
+	additionalInfo := m[CtEventAdditionalInfo]
+	if additionalInfo == "" {
```
[the remark this answers is not in the document]
Comment (fperot74) [reply database/events.go]: Done by Sonia in the bridge... This way, it remains compatible with her version and new developments but you have to choose to set additional_info by your self OR let this function set it

File: database/events.go [hunk 5]
```diff
@@ -48,41 +96,54 @@ func NewEventsDBModule(db CloudtrustDB) EventsDBModule {
...
+		var addNfo = make(map[string]string)
```
[the remark this answers is not in the document]
Comment (fperot74) [reply database/events.go]: I often write Nfo for information... or Info... yes, I'm a little bit lazy

**Pull request merged**